pkg/sql/backfill/backfill.go

```diff
@@ -372,36 +380,142 @@ func (ib *IndexBackfiller) ContainsInvertedIndex() bool {
 	return false
 }
 
-// Init initializes an IndexBackfiller.
-func (ib *IndexBackfiller) Init(
-	evalCtx *tree.EvalContext, desc *sqlbase.ImmutableTableDescriptor,
+// InitForLocalUse initializes an IndexBackfiller for use during local execution
+// within a transaction. In this case, the entire backfill process is occurring
+// on the gateway as part of the user's transaction.
+func (ib *IndexBackfiller) InitForLocalUse(
+	ctx context.Context,
+	evalCtx *tree.EvalContext,
+	semaCtx *tree.SemaContext,
+	desc *sqlbase.ImmutableTableDescriptor,
 ) error {
-	ib.evalCtx = evalCtx
-	numCols := len(desc.Columns)
-	cols := desc.Columns
+	// Initialize ib.cols and ib.colIdxMap.
+	ib.initCols(desc)
+
+	// Initialize ib.added.
+	valNeededForCol := ib.initIndexes(desc)
+
+	// Convert any partial index predicate strings into expressions.
+	predicates, predicateRefColIDs, err := schemaexpr.MakePartialIndexExprs(
+		ctx,
+		ib.added,
+		ib.cols,
+		desc,
+		evalCtx,
+		semaCtx,
+	)
+	if err != nil {
+		return err
+	}
+
+	// Add the columns referenced in the predicate to valNeededForCol so that
+	// columns necessary to evaluate the predicate expression are fetched.
+	predicateRefColIDs.ForEach(func(col sqlbase.ColumnID) {
+		valNeededForCol.Add(ib.colIdxMap[col])
+	})
+
+	return ib.init(evalCtx, predicates, valNeededForCol, desc)
+}
+
+// InitForDistributedUse initializes an IndexBackfiller for use as part of a
+// backfill operation executing as part of a distributed flow. In this use, the
+// backfill operation manages its own transactions. This separation is necessary
+// due to the different procedure for accessing user defined type metadata as
+// part of a distributed flow.
+func (ib *IndexBackfiller) InitForDistributedUse(
+	ctx context.Context, flowCtx *execinfra.FlowCtx, desc *sqlbase.ImmutableTableDescriptor,
+) error {
+	// Initialize ib.cols and ib.colIdxMap.
+	ib.initCols(desc)
+
+	// Initialize ib.added.
+	valNeededForCol := ib.initIndexes(desc)
+
+	evalCtx := flowCtx.NewEvalCtx()
+	var predicates map[sqlbase.IndexID]tree.TypedExpr
+	var predicateRefColIDs sqlbase.TableColSet
+
+	// Install type metadata in the target descriptors, as well as resolve any
+	// user defined types in the column expressions.
```

Comment:
[nit] any user defined types in partial index expressions

Reply:
Good catch. Done.